core/builtin.py

```diff
@@ -235,6 +236,84 @@ def Echo(argv):
   return 0
 
 
+PRINTF_SPEC = _Register('printf')
+PRINTF_SPEC.ShortFlag('-v', args.Str)
+
+
+def Printf(argv, mem):
+  """
+  printf: printf [-v var] format [argument ...]
+  """
+  arg, args_consumed = PRINTF_SPEC.Parse(argv)
+  if args_consumed >= len(argv):
+    util.error('printf: need format string')
+    return 1
+  fmt = argv[args_consumed]
+  vals = argv[args_consumed + 1:]
+
+  result = ''
+  f = 0
+  v = 0
+  # Loop invariant: vals[v:] and fmt[f:] remain, to accumulate into result.
+  while True:
+    f_next = fmt.find('%', f)
+    if f_next < 0:
+      f_next = len(fmt)
+
+    result += fmt[f:f_next]  # TODO backslash-escapes, at least \n
+    f = f_next
+
+    if f >= len(fmt):
+      if v >= len(vals):
+        break
+      else:
+        # (handy!) bash printf quirk: re-use fmt to consume remaining vals.
+        f = 0
+        continue
+
+    c = fmt[f+1]
+    if c == '%':
+      f += 2
+      result += '%'
```

Comment:
Please use the `parts.append` and `''.join(parts)` idiom here.  I know CPython optimizes it not to be quadratic, but I already started writing the OVM interpreter loop and it won't be special-cased :)

Reply:
OK :)